sdks/java/io/kinesis2/src/test/java/org/apache/beam/sdk/io/kinesis2/KinesisIOIT.java

```diff
@@ -0,0 +1,155 @@
+
+import com.amazonaws.regions.Regions;
+import java.io.Serializable;
+import java.nio.charset.StandardCharsets;
+import java.util.Random;
+import org.apache.beam.sdk.io.GenerateSequence;
+import org.apache.beam.sdk.io.common.HashingFn;
+import org.apache.beam.sdk.io.common.TestRow;
+import org.apache.beam.sdk.io.kinesis.KinesisPartitioner;
+import org.apache.beam.sdk.options.PipelineOptionsFactory;
+import org.apache.beam.sdk.testing.PAssert;
+import org.apache.beam.sdk.testing.TestPipeline;
+import org.apache.beam.sdk.transforms.Combine;
+import org.apache.beam.sdk.transforms.Count;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.values.PCollection;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+import org.junit.BeforeClass;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.kinesis.common.InitialPositionInStream;
+
+/**
+ * Integration test, that writes and reads data to and from real Kinesis. You need to provide {@link
+ * KinesisTestOptions} in order to run this.
+ */
+@RunWith(JUnit4.class)
+public class KinesisIOIT implements Serializable {
+  private static int numberOfShards;
+  private static int numberOfRows;
+
+  @Rule public TestPipeline pipelineWrite = TestPipeline.create();
+  @Rule public TestPipeline pipelineRead = TestPipeline.create();
+
+  private static KinesisTestOptions options;
+  private static final Instant now = Instant.now();
+
+  @BeforeClass
+  public static void setup() {
+    PipelineOptionsFactory.register(KinesisTestOptions.class);
+    options = TestPipeline.testingPipelineOptions().as(KinesisTestOptions.class);
+    numberOfShards = options.getNumberOfShards();
+    numberOfRows = options.getNumberOfRecords();
+  }
+
+  /** Test which write and then read data for a Kinesis stream. */
+  @Test
+  public void testWriteThenRead() {
+    runWrite();
+    runRead();
+  }
+
+  /** Write test dataset into Kinesis stream. */
+  private void runWrite() {
+    pipelineWrite
+        .apply("Generate Sequence", GenerateSequence.from(0).to((long) numberOfRows))
+        .apply("Prepare TestRows", ParDo.of(new TestRow.DeterministicallyConstructTestRowFn()))
+        .apply("Prepare Kinesis input records", ParDo.of(new ConvertToBytes()))
+        .apply(
+            "Write to Kinesis",
+            org.apache.beam.sdk.io.kinesis.KinesisIO.write()
+                .withStreamName(options.getAwsKinesisStream())
+                .withPartitioner(new RandomPartitioner())
+                .withAWSClientsProvider(
+                    options.getAwsAccessKey(),
+                    options.getAwsSecretKey(),
+                    Regions.fromName(options.getAwsKinesisRegion())));
+
+    pipelineWrite.run().waitUntilFinish();
+  }
+
+  /** Read test dataset from Kinesis stream. */
+  private void runRead() {
```

Comment:
Since we can read messaged using two different modes - pull and subscribe - then I think we need to have two different Read tests for that.